File: src/main/java/com/example/application/views/updates/SoftwareUpdateFormSolution.java
package com.example.application.views.updates;

import com.example.application.data.entity.SoftwareUpdate;
import com.example.application.data.entity.TeslaModel;
import com.example.application.data.service.TeslaModelService;
import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.combobox.MultiSelectComboBox;
import com.vaadin.flow.component.datepicker.DatePicker;
import com.vaadin.flow.component.formlayout.FormLayout;
import com.vaadin.flow.component.textfield.TextArea;
import com.vaadin.flow.component.textfield.TextField;
import com.vaadin.flow.data.binder.BeanValidationBinder;
import com.vaadin.flow.data.binder.ValidationException;

import java.util.HashSet;
import java.util.stream.Collectors;

public class SoftwareUpdateFormSolution extends FormLayout {

    private SoftwareUpdate editedSoftwareUpdate = new SoftwareUpdate();
    private final BeanValidationBinder<SoftwareUpdate> binder;

    public SoftwareUpdateFormSolution(TeslaModelService teslaModelService) {
        TextField version = new TextField("Version");
        DatePicker releaseDate = new DatePicker("Release Date");
        add(version, releaseDate);

        binder = new BeanValidationBinder<>(SoftwareUpdate.class);
        binder.forField(version).bind(SoftwareUpdate::getVersion, SoftwareUpdate::setVersion);
        binder.forField(releaseDate).bind(SoftwareUpdate::getReleaseDate, SoftwareUpdate::setReleaseDate);

        /* Task 1:
         * Allow users to select which car models an update should be applied to, using a multi-select combo box.
         * Users must select at least one model.
         * You can use the TeslaModelService to get a list of all available models.
         */
        MultiSelectComboBox<TeslaModel> models = new MultiSelectComboBox<>("Models");
        models.setItemLabelGenerator(TeslaModel::getName);
        models.setItems(teslaModelService.listAll());
        add(models);

        binder.forField(models).asRequired().bind(SoftwareUpdate::getModels, SoftwareUpdate::setModels);

        /* Task 2:
         * Display the currently selected models in a text box (for example a read-only text area) at the end of the form.
         */
        TextArea selectedModels = new TextArea("Selected models");
        selectedModels.setReadOnly(true);
        add(selectedModels);
        models.addValueChangeListener(e -> {
           String modelsText = e.getValue().stream().map(TeslaModel::getName).collect(Collectors.joining(","));
           selectedModels.setValue(modelsText);
        });

        /* Task 3:
         * Add a button that selects all models.
         * The button should only be enabled if no model is selected yet.
         */
        Button selectAll = new Button("Select all", e -> {
            models.setValue(new HashSet<>(teslaModelService.listAll()));
        });
        add(selectAll);
        models.addValueChangeListener(e -> {
            boolean hasEmptyValue = e.getValue().isEmpty();
            selectAll.setEnabled(hasEmptyValue);
        });
    }

    public void edit(SoftwareUpdate softwareUpdate) {
        this.editedSoftwareUpdate = softwareUpdate;
        binder.readBean(this.editedSoftwareUpdate);
    }

    public SoftwareUpdate submit() throws ValidationException {
        binder.writeBean(this.editedSoftwareUpdate);
        return this.editedSoftwareUpdate;
    }

    public void clear() {
        this.edit(new SoftwareUpdate());
    }
}

File: frontend/views/updateshilla/software-update-form-solution.ts
import { html, LitElement } from 'lit';
import { customElement } from 'lit/decorators.js';
import '@vaadin/button';
import '@vaadin/date-picker';
import '@vaadin/form-layout';
import '@vaadin/multi-select-combo-box';
import { MultiSelectComboBoxSelectedItemsChangedEvent } from '@vaadin/multi-select-combo-box';
import '@vaadin/text-area';
import '@vaadin/text-field';
import SoftwareUpdate from 'Frontend/generated/com/example/application/data/entity/SoftwareUpdate';
import { state } from 'lit/decorators';
import TeslaModel from 'Frontend/generated/com/example/application/data/entity/TeslaModel';
import { SoftwareUpdateEndpoint, TeslaModelEndpoint } from 'Frontend/generated/endpoints';
import { TextFieldValueChangedEvent } from '@vaadin/text-field';
import { DatePickerValueChangedEvent } from '@vaadin/date-picker';

/* Task 1:
 * Allow users to select which car models an update should be applied to, using a multi-select combo box.
 * You can use the availableModels property to get a list of all availabe models.
 */

/* Task 2:
 * Display the currently selected models in a text box (for example a read-only text area) at the end of the form.
 */

/* Task 3:
 * Add a button that selects all models.
 * The button should only be enabled if no model is selected yet.
 */

@customElement('software-update-form-solution')
export class SoftwareUpdateFormSolution extends LitElement {
  @state()
  private editedSoftwareUpdate!: SoftwareUpdate;

  @state()
  private availableModels: TeslaModel[] = [];

  get selectedModelsText(): string {
    return this.editedSoftwareUpdate.models.map((model) => model.name).join(', ');
  }

  get isSelectAllModelsEnabled(): boolean {
    return this.editedSoftwareUpdate.models.length === 0;
  }

  async connectedCallback() {
    super.connectedCallback();
    this.clear();
    this.availableModels = await TeslaModelEndpoint.listAll();
  }

  edit(softwareUpdate: SoftwareUpdate) {
    this.editedSoftwareUpdate = softwareUpdate;
  }

  async submit() {
    return SoftwareUpdateEndpoint.update(this.editedSoftwareUpdate);
  }

  clear() {
    this.editedSoftwareUpdate = {
      version: '',
      releaseDate: '',
      models: [],
    };
  }

  render() {
    return html`
      <vaadin-form-layout>
        <vaadin-text-field
          label="Version"
          id="version"
          .value="${this.editedSoftwareUpdate.version}"
          @value-changed="${(e: TextFieldValueChangedEvent) =>
            (this.editedSoftwareUpdate = {
              ...this.editedSoftwareUpdate,
              version: e.detail.value,
            })}"
        ></vaadin-text-field>
        <vaadin-date-picker
          label="Release Date"
          id="lastName"
          .value="${this.editedSoftwareUpdate.releaseDate}"
          @value-changed="${(e: DatePickerValueChangedEvent) =>
            (this.editedSoftwareUpdate = {
              ...this.editedSoftwareUpdate,
              releaseDate: e.detail.value,
            })}"
        ></vaadin-date-picker>
        <vaadin-multi-select-combo-box
          label="Models"
          item-id-path="id"
          item-label-path="name"
          .items="${this.availableModels}"
          .selectedItems="${this.editedSoftwareUpdate.models}"
          @selected-items-changed="${(e: MultiSelectComboBoxSelectedItemsChangedEvent<TeslaModel>) =>
            (this.editedSoftwareUpdate = {
              ...this.editedSoftwareUpdate,
              models: e.detail.value,
            })}"
        ></vaadin-multi-select-combo-box>
        <vaadin-text-area label="Selected models" readonly .value="${this.selectedModelsText}"></vaadin-text-area>
        <vaadin-button
          label="Selected models"
          .disabled="${!this.isSelectAllModelsEnabled}"
          .value="${this.selectedModelsText}"
          @click="${(e: Event) =>
            (this.editedSoftwareUpdate = {
              ...this.editedSoftwareUpdate,
              models: [...this.availableModels],
            })}"
          >Select all
        </vaadin-button>
      </vaadin-form-layout>
    `;
  }
}
